lib/linters.js

```diff
@@ -121,9 +121,140 @@ const js = {
+      piece type, add a second piece type for that purpose and create a "relationship"
+      field to connect them.
+    `
+  },
+  joinByOne: {
+    match: /joinByOne/g,
+    message: stripIndent`
+      Apostrophe A3 has no "joinByOne" field type. Use the "relationship" field
+      type instead and set both "required: true" and "max: 1". The loaded data will
+      appear as an array with a single element.
+    `
+  },
+  joinByOneReverse: {
+    match: /joinByOneReverse/g,
+    message: stripIndent`
+      Apostrophe A3 has no "joinByOneReverse" field type. Change your "joinByOne"
+      field to a "relationship" field as noted elsewhere, then change "joinByOneReverse"
+      to "relationshipReverse".
+    `
+  },
+  joinByArray: {
+    match: /joinByArray/g,
+    message: stripIndent`
+      In A3 the "joinByArray" field type has been renamed to "relationship".
+    `
+  },
+  joinByArrayReverse: {
+    match: /joinByArrayReverse/g,
+    message: stripIndent`
+      Apostrophe A3 has no "joinByArrayReverse" field type. Change your "joinByArray"
+      field to a "relationship" field as noted elsewhere, then change "joinByArrayReverse"
+      to "relationshipReverse".
+    `
+  },
+  tasks: {
+    match: /apos\.tasks\.add|self\.addTask/g,
+    message: stripIndent`
+      Apostrophe tasks have moved to the new "tasks" section.
+    `,
+    url: 'https://v3.docs.apostrophecms.org/reference/module-api/module-overview.html#tasks'
+  },
+  widgetOutput: {
+    match: /self\.output\s*=/g,
+    message: stripIndent`
+      Overriding the "output" method of a widget module is usually unnecessary
+      as it simply renders the template. If you must override it be aware
+      that the arguments have changed to:
+      
+      "req, widget, options, _with"
+      
+      Also, the output method is now async.
+    `
+  },
+  filterOptionsForDataAttribute: {
+    match: /filterOptionsForDataAttribute/g,
+    message: stripIndent`
+      filterOptionsForDataAttribute is not needed in A3, as A3 does not add any data
+      attributes to a widget by default, except when editing. Your widget templates
+      are responsible for making any needed data available to widget players via data
+      attributes.
+    `
+  },
+  addFieldType: {
+    match: /(schema|self)\.addFieldType/g,
+    message: stripIndent`
+      Be aware that addFieldType has changed in A3. There is just one "convert"
+      method per field type, which should accept either a string representation
+      for import purposes (CSV, etc) and may also accept a different representation
+      for form submissions where appropriate. "convert" may be async and now receives
+      these arguments, typically copying from "data[field.name]" to
+      "destination[field.name]" after sanitization:
+
+      "req, field, data, destination"
+    `
+  },
+  joinByArraySchema: {
+    match: /(joinByArray|relationship)[\s\S]{1,1000}schema["']?:/,
```

Comment:
Compromise technique to find schema "pretty close" to the field type, as we're not doing strict ecmascript parsing here.

Reply:
This worked well in some modules and projects tested.